polkadot/node/network/collator-protocol/src/validator_side/collation.rs

```diff
@@ -267,33 +334,179 @@ impl Collations {
+				max_candidate_depth: _,
+				allowed_ancestry_len: _,
+			} => {
+				let pending_for_para = self.pending_for_para(para_id);
+
+				let respected_per_para_limit =
+					self.claims_per_para.get(&para_id).copied().unwrap_or_default() >
+						seconded_for_para + pending_for_para;
+
+				gum::trace!(
+					target: LOG_TARGET,
+					?para_id,
+					claims_per_para=?self.claims_per_para,
+					seconded_per_para=?self.seconded_per_para,
+					?pending_for_para,
+					?respected_per_para_limit,
+					"is_seconded_limit_reached - ProspectiveParachainsMode::Enabled with claim queue support"
+				);
+
+				!respected_per_para_limit
+			},
+		}
+	}
+
+	/// Adds a new collation to the waiting queue for the relay parent. This function doesn't
+	/// perform any limits check. The caller (`enqueue_collation`) should assure that the collation
+	/// limit is respected.
+	pub(super) fn add_to_waiting_queue(&mut self, collation: (PendingCollation, CollatorId)) {
+		self.waiting_queue.entry(collation.0.para_id).or_default().push_back(collation);
+	}
+
+	/// Picks a collation to fetch from the waiting queue.
+	/// When fetching collations we need to ensure that each parachain has got a fair core time
+	/// share depending on its assignments in the claim queue. This means that the number of
+	/// collations seconded per parachain should ideally be equal to the number of claims for the
+	/// particular parachain in the claim queue.
+	///
+	/// To achieve this each seconded collation is mapped to an entry from the claim queue. The next
+	/// fetch is the first unsatisfied entry from the claim queue for which there is an
+	/// advertisement.
+	///
+	/// If claim queue is not supported then `group_assignment` should contain just one element and
+	/// the score won't matter. In this case collations will be fetched in the order they were
+	/// received.
+	///
+	/// Note: `group_assignments` is needed just for the fall back logic. It should be removed once
+	/// claim queue runtime api is released everywhere since it will be redundant - claim queue will
+	/// already be available in `self.claim_queue_state`.
+	fn pick_a_collation_to_fetch(
+		&mut self,
+		group_assignments: &Vec<ParaId>,
+	) -> Option<(PendingCollation, CollatorId)> {
+		gum::trace!(
+			target: LOG_TARGET,
+			waiting_queue=?self.waiting_queue,
+			claims_per_para=?self.claims_per_para,
+			?group_assignments,
+			"Pick a collation to fetch."
+		);
+
+		let claim_queue_state = match self.claim_queue_state.as_mut() {
+			Some(cqs) => cqs,
+			// Fallback if claim queue is not available. There is only one assignment in
+			// `group_assignments` so fetch the first advertisement for it and return.
+			None =>
+				if let Some(assigned_para_id) = group_assignments.first() {
+					return self
+						.waiting_queue
+						.get_mut(assigned_para_id)
+						.and_then(|collations| collations.pop_front())
+				} else {
+					unreachable!("Group assignments should contain at least one element.")
```

Comment:
How can this be true? What if nothing is scheduled on the core?

Reply:
This has been reworked.